### PR TITLE
ARROW-75: Fix handling of empty strings

### PR DESCRIPTION
**Pull request by danrobinson**
Fixes [ARROW-75](https://issues.apache.org/jira/browse/ARROW-75) (and changes Python tests to verify that behavior).

### CODE REVIEW

File: cpp/src/arrow/types/primitive.h [hunk 1]
```diff
@@ -168,7 +168,7 @@ class PrimitiveBuilder : public ArrayBuilder {
       int32_t new_capacity = util::next_power2(length_ + length);
       RETURN_NOT_OK(Resize(new_capacity));
     }
-    memcpy(raw_buffer() + length_, values, length * elsize_);
+    if (length > 0) memcpy(raw_buffer() + length_, values, length * elsize_);
```
Comment (wesm) [inline cpp/src/arrow/types/primitive.h]: Put braces around this (with the memcpy on its own line)?

**Comment by wesm** [comment]
Could you also add a small unit test to `types/string-test.cc` (appending one or two empty strings and checking the resulting `StringArray` will suffice)?

**Comment by danrobinson** [comment]
It turns out the empty strings only cause a segfault when they are the first item in the array. Both the Python and C++ tests now test for this behavior.

**Comment by wesm** [comment]
You can add a pre-commit hook to run `make lint` to avoid cpplint-related travis failures (or also run `make lint` -- I forgot a lot until I added the pre-commit hook)

**Comment by wesm** [comment]
+1, thank you

**Comment by wesm** [comment]
Merged -- note you should be able to self-assign JIRAs now

**Pull request closed without merging**